### DISCUSSION TITLE
Intended for Windows use?

### DISCUSSION THREAD

**Discussion by Aptronymist** [vladmandic/sdnext#71] [Q&A] [answered]
I'm quite interested in trying this out (which I now have in cpu mode), but after cloning it and installing the requirements I noticed that there was no webui-user.bat file. 
I just copied the one I'm using over from standard webui, tossed in the --skip-torch-cuda-test (AMD user sadly) and fired it up, other than being stuck in cpu mode it's okay. I'm guessing since you're working at a rapid pace you may have forgotten about that batch file, or you didn't intend it for windows, either way.

Any chance of this getting any DirectML/Vulkan/Onnx functionality in from other forks like SHARK? AMD support is woefully inadequate atm and you seem to be quite the trailblazer.

**Comment by vladmandic** [comment, marked as answer]
you're correct in the assumption. but i'm not sure if adding the file that is used only to set env variables is a way to go?  
perhaps a better way would be to auto-detect amd vs nvidia and setup correct environment automatically?  
which env variables are you currently using?

regarding adding directml/vulkan/onnx functionality, i'll take a look at that if its feasible - but it would also heavily depend on ability to test on nvidia as i don't have an amd card and i really don't want to be adding code blindly.

**Reply by Aptronymist** [reply]
True, it probably wasn't the best option, it was simply a quick and dirty method and seemed to work okay, especially since I don't think any of it would really matter in cpu mode. I really liked your UI changes, it's very clean. honestly I only did it to set the cuda bypass flag to get it going in cpu mode.

**Comment by vladmandic** [comment]
btw, i'm not in favor of bringing in support for `onnx` and `tensorrt` since they both require model conversion - they cannot work with existing large model galleries such as ones on <https://civitai.com/>

on the other hand, `directml` is feasible - if there is sufficient demand. i took a look at some examples and its a non-trivial work since how its achieved is by patching diffuser repo directly instead of doing in it runtime. so its either just directml or no directml at all - i'd need to make it pluggable so directml becomes one option and not the only option :)

**Reply by Aptronymist** [reply]
I am currently using the [DirectML AMD specific fork](https://github.com/lshqqytiger/stable-diffusion-webui-directml), like most others AMD users not running linux I assume, and it's far from perfect but seems to do okay most of the time, though I do find numerous things don't work right (or at all). From what I can tell it's just applying some wide-ranging DML patches to the existing code. The performance isn't great but it does work.
iirc [SHARK ](https://github.com/nod-ai/SHARK) is using Vulkan and torch-MLIR/IREE, though it does seem to do some kind of conversion, I haven't messed with it much, so I'm not sure if that's necessary.

Your attention would be greatly appreciated I'm sure, even the replies you have given are open and welcoming, which I appreciate. I'm trying to scrounge together the cash for an nvidia card, but they're pretty outrageous atm.
If you specifically need access to an rx590 8gb, I'd be happy to provide it as best I could. Linux on WSL2 or even a quick install via ventoy (I run bare metal), I'm sure others would be happy to do the same if they knew someone would stop pretending they didn't exist! 

Thank you for your consideration, and I look forward to your future work regardless, your _opinions_ look great to me.

**Reply by vladmandic** [reply]
thanks for the comments!

and yes, that's exactly the fork i was looking at - the actual webui code changes from there can all be ported, but biggest issue is that it patches some things directly into compvis-stable-diffusion and k-diffusion repos which is a no-no, i'd need to extract them and make them loadable instead. and from an initial look, i only see `k-diffusers` as patched, so some that don't fall into k-diffusors (like my new favorite - `unipc`) are probably broken.

shark looks very interesting and i'll definitely give it a closer look, but from a mile-high look, its veery different, so not sure if its portable.

regarding, skip cuda test, i actually don't see benefit of erroring out if not detected - i think i'll change it warning so flag wont be needed.

**Reply by Aptronymist** [reply]
> thanks for the comments!
> 
> and yes, that's exactly the fork i was looking at - the actual webui code changes from there can all be ported, but biggest issue is that it patches some things directly into compvis-stable-diffusion and k-diffusion repos which is a no-no, i'd need to extract them and make them loadable instead. and from an initial look, i only see `k-diffusers` as patched, so some that don't fall into k-diffusors (like my new favorite - `unipc`) are probably broken.

I'm trying to remember, but I don't think I've ever had unipc work correctly, I tried it a few times just now and it's not a good result.

**Reply by vladmandic** [reply]
Yeah, thats what I expected based on quick look of the code.

**Reply by Aptronymist** [reply]
The new ToMe also doesn't seem to work either, at least for me, though admittedly I have made a few minor alterations myself that might contribute (like inserting the Negative Guidance minimum sigma PR code I saw your comments on into the DML fork), as well as several other extensions that I just didn't bother keeping track of.

**Comment by Lumpsack** [comment]
Would NCNN be an option as used by ChainNer (https://github.com/chaiNNer-org/chaiNNer) ?
Apologies, I don't know a massive amount about how it works or embeds, or if its even compatible in any way, but as I found it recently it could be you're not aware of it either.
I think this is NCNN (https://github.com/Tencent/ncnn) and looks to be tied to vulkan from my quick skim over - but perhaps a simpler route?
Anyway I really appreciate you considering AMD users and for your time.

**Reply by vladmandic** [reply]
ncnn is a completely different framework, it doesn't work with torch at all, so it would be rewrite-from-scratch.
for amd users on windows, best option is to hope that amd releases rocm 5.5 for windows (its in alpha now). as soon as its released, i can probably support it pretty fast.

**Reply by Lumpsack** [reply]
Thats good to know and thank you, fingers crossed it'll happen in my lifetime ;)
